### PR TITLE
ADD: Logistic Regression and TCN support

### PR DESCRIPTION
**Pull request by matthewbayer**

### CODE REVIEW

**Comment by madisonmay** [comment]
@matthewbayer I think you're missing a few files in this commit :P

File: finetune/base_models/tcn/featurizer.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+                    name="conv1_" + str(layer_num),
+                )
+                self.conv2 = tf.keras.layers.Conv1D(
+                    filters=n_filters,
+                    kernel_size=kernel_size,
+                    padding="same",
+                    dilation_rate=dilation_rate,
+                    activation=tf.nn.relu,
+                    kernel_initializer=tf.initializers.glorot_normal,
+                    name="conv1_" + str(layer_num),
```
Comment (benleetownsend) [inline finetune/base_models/tcn/featurizer.py]: this seems wrong that both convs have the same name

File: finetune/base_models/tcn/featurizer.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    encoder,
+    config,
+    train=False,
+    reuse=None,
+    context=None,
+    context_dim=None,
+    **kwargs
+):
+    """
+    The transformer element of the finetuning model. Maps from tokens ids to a dense, embedding of the sequence.
```
Comment (lhz1029) [inline finetune/base_models/tcn/featurizer.py]: You probably want to update the docs?

File: finetune/base_models/tcn/featurizer.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+        # mask is past the classify token (i.e. make those results extremely negative)
+        mask = tf.expand_dims(
+            1.0
+            - tf.sequence_mask(
+                pool_idx, maxlen=tf.shape(representation)[1], dtype=tf.float32
+            ),
+            -1,
+        )
+        pool = tf.reduce_max(representation + mask * -1e9, 1)
```
Comment (lhz1029) [inline finetune/base_models/tcn/featurizer.py]: The masking and pooling is the same as for the textcnn. Could you pull these out to a shared function?

File: finetune/base_models/tcn/featurizer.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    X = tf.reshape(X, shape=tf.concat(([-1], initial_shape[-2:]), 0))
+
+    with tf.variable_scope("model/featurizer", reuse=reuse):
+        embed_weights = tf.get_variable(
+            name="we",
+            shape=[encoder.vocab_size + config.max_length, config.n_embed_featurizer],
+            initializer=tf.random_normal_initializer(stddev=config.weight_stddev),
+        )
+
+        if config.train_embeddings:
```
Comment (madisonmay) [inline finetune/base_models/tcn/featurizer.py]: We should make sure to set config.train_embeddings to false for this model type

File: finetune/base_models/tcn/model.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+from finetune.base_models.gpt2.encoder import GPT2Encoder
+from finetune.base_models.tcn.featurizer import tcn_featurizer
+from finetune.util.download import GPT2_BASE_URL, FINETUNE_BASE_FOLDER
+
+
+class TCNModel(SourceModel):
+    is_bidirectional = False
+    encoder = GPT2Encoder
+    featurizer = tcn_featurizer
+    settings = {
```
Comment (madisonmay) [inline finetune/base_models/tcn/model.py]: Important that we set n_epochs to a higher value for TCNModel since we initialize from random.  Maybe 5 to start?

File: finetune/target_models/logistic_regression.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    :param config: A :py:class:`finetune.config.Settings` object or None (for default config).
+    :param \**kwargs: key-value pairs of config items to override.
+    """
+    defaults = {'encoder': 'glove'}
+    
+
+    def __init__(self, **kwargs):
+        self.LR = LogisticRegression(penalty="l2", max_iter=500, solver="lbfgs", multi_class='multinomial')
+        c = kwargs.pop('C', None)
+        if c: self.LR.C = c
```
Comment (madisonmay) [inline finetune/target_models/logistic_regression.py]: So many if statement one liners.

I'm convinced this was intended to be a personal attack.

Comment (matthewbayer) [reply finetune/target_models/logistic_regression.py]: if madison_is_angry: keep_doing_it()

File: finetune/base_models/tcn/featurizer.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+                # residuals
+                if X.get_shape()[-1] == self.n_filters:
+                    output = conv2_dropout + X
+                else:
+                    downsampled_out = self.downsample(X)
+                    output = conv2_dropout + downsampled_out
+
+                return output
+
+        with tf.variable_scope("crf"):
```
Comment (benleetownsend) [inline finetune/base_models/tcn/featurizer.py]: Can we change this scope name? Its kinda confusing because we don't have the crf in here.

File: finetune/saver.py [hunk 8]
```diff
@@ -238,6 +240,7 @@ def init_fn(scaffold, session, model_portion=None, refresh_base_model=False):
             else:
                 variables_sv = dict()
             all_vars = tf.global_variables()
+            print(all_vars)
```
Comment (benleetownsend) [inline finetune/saver.py]: Oops

File: finetune/base_models/tcn/featurizer.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+        X = tf.reshape(X, [-1, config.max_length, 2])
+
+        # we remove positional embeddings from the model
+        h = embed(X[:, :, :1], embed_weights)
+
+        # keep track of the classify token
+        clf_token = encoder["_classify_"]
+        
+        class TemporalBlock:
```
Comment (madisonmay) [inline finetune/base_models/tcn/featurizer.py]: Shouldn't be inlined, gonna remove.
